data_steward/bq_utils.py

```diff
@@ -511,3 +546,69 @@ def _transform_row(row, schema):
 def _list_all_table_ids(dataset_id):
     tables = list_tables(dataset_id)
     return [table['tableReference']['tableId'] for table in tables]
+
+
+def create_dataset(
+        project_id=None,
+        dataset_id=None,
+        description=None,
+        friendly_name=None,
+        overwrite_existing=None
+    ):
+    """
+    Creates a new dataset from the API.
+
+    :param  project_id:  name of the project to create the dataset in.  defaults
+        to the currently configured project if missing.
+    :param  dataset_id:  name to give the new dataset.  it is required.
+    :param  description:  dataset description.  it is required.
+    :param  friendly_name:  an user friendly name for the dataset.  optional.
+    :param  overwrite_existing:  Determine if the dataset should be overwritten
+        if it already exists.  defaults to true / overwrite.
+
+    """
+    if dataset_id is None:
+        raise RuntimeError("Cannot create a dataset without a name")
+
+    if description is None:
+        raise RuntimeError("Will not create a dataset without a description")
+
+    if project_id is None:
+        app_id = app_identity.get_application_id()
+    else:
+        app_id = project_id
+
+    if overwrite_existing is None:
+        overwrite_existing = bq_consts.TRUE
+    else:
+        overwrite_existing = bq_consts.FALSE
+
+    bq_service = create_service()
+
+    job_body = {
+        bq_consts.DATASET_REFERENCE: {
+            bq_consts.PROJECT_ID: app_id,
+            bq_consts.DATASET_ID: dataset_id,
+        },
+        bq_consts.DESCRIPTION: description,
+    }
+
+    if friendly_name:
+        job_body.update({bq_consts.FRIENDLY_NAME: friendly_name})
+
+    insert_dataset = bq_service.datasets().insert(projectId=app_id, body=job_body)
+    try:
+        insert_result = insert_dataset.execute(num_retries=bq_consts.BQ_DEFAULT_RETRY_COUNT)
+    except HttpError:
```

Comment:
We wanted to make sure that the HTTP error is raised only if the dataset is already available in the project, because we should not be trying to delete the dataset if the HTTP error is due to other reason.

Reply:
Updated the code to log when it creates the dataset successfully, log when it deletes an existing dataset and creates a new one, log/raise when it tries to delete an existing dataset but the overwrite parameter is set to false so it fails, and log/raise when any other HttpError is caught.